File: 230. Kth Smallest Element in a BST/main.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func main() {
	ans := kthSmallest(&TreeNode{
		Val: 3,
		Left: &TreeNode{
			Val: 1,
			Right: &TreeNode{
				Val: 2,
			},
		},
		Right: &TreeNode{
			Val: 4,
		},
	}, 1)
	fmt.Println(ans)
}

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

var rank int
var res int

func kthSmallest(root *TreeNode, k int) int {
	rank = 0
	res = 0
	traverse(root, k)
	return res
}

func traverse(root *TreeNode, k int) {
	if root == nil {
		return
	}

	traverse(root.Left, k)
	rank++
	if rank == k {
		res = root.Val
		return
	}
	traverse(root.Right, k)
}

File: 230. Kth Smallest Element in a BST/main.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        pos = 0
        res = 0
        def traverse(root: Optional[TreeNode], k: int):
            if not root:
                return
            
            traverse(root.left, k)

            nonlocal pos
            pos += 1
            if k == pos:
                nonlocal res
                res = root.val
                return

            traverse(root.right, k)
        traverse(root, k)
        return res
